### DISCUSSION TITLE
Do the data normalization, feature selection, and SMOTE in pycaret cause data leakage?

### DISCUSSION THREAD

**Discussion by Sepehr-76** [pycaret/pycaret#3755] [Q&A]
Hello! I am new to pycaret and I was wondering if the automatic transformations in the setup such as data normalization, feature selection, and SMOTE are performed separately in the train and test data sets or are they performed on all of the the data? In the second case there is worry for data leakage. Thanks for the help

**Comment by tvdboom** [comment]
The transformation pipeline is fitted on the training set only. Transformations are then done on the complete dataset. Note that some transformers like SMOTE are only applied to the training set since you don't want to do class balancing on the test set.

There should be no data leakage, but if you have strong evidence there is, please open an issue.

**Reply by TomLucidor** [reply]
It's one of those textbook things where the concept itself need some extra elaboration: the test set is only transformed based on the distribution (or data characteristic) of the training set, not the whole dataset. Therefore, there is no possible leakage. "Test-time training" does not really exist in this case.
